File: c/totp.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include <openssl/sha.h>

// Sleep
#include <unistd.h>
#include <time.h>

#define sha_block_size SHA_CBLOCK

#define OPAD "kkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkkk"
#define IPAD "mmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm"

int max(int a, int b){
    if(a > b) {
        return a;
    } else {
        return b;
    }
}

void str_xor(char* c, int lenA, char* a, int lenB, char* b) {

    //printf("A: %s; B: %s; Len A: %d; Len B: %d;\n", a, b, lenA, lenB);
    int lenC = max(lenA, lenB);

    char* aaux = malloc(lenC);
    char* baux = malloc(lenC);

    int i;
    for(i = lenC; i >= 0; i--){

        int itAux = lenC - i;

        if(lenA < i){
            aaux[itAux] = '\x00';
        } else {
            aaux[itAux] = a[itAux];
        }

        if(lenB < i){
            baux[itAux] = '\x00';
        } else {
            baux[itAux] = b[itAux];
        }
    }

    for(i = 0; i < lenC; i++)
        *(c + i) = aaux[i] ^ baux[i];

    free(aaux);
    free(baux);
}

void hmac(char* res, int klen, char* K, int mlen, char* m){

    char* k = malloc(SHA_DIGEST_LENGTH);

    if(klen > sha_block_size){
        SHA1((const unsigned char*) K, klen, (unsigned char*) k);
        klen = SHA_DIGEST_LENGTH;
    } else {
      strcpy(k, K);
    }

    if(klen < sha_block_size) {
        char aux[sha_block_size];

        int s = klen;
        int i;
        for(i = 0; i < sha_block_size; i++) {
            if (sha_block_size - i > s) {
                aux[i] = '\x00';
            } else {
                aux[i] = k[i - sha_block_size + s];
            }
        }

        k = aux;

    }

    char hash[SHA_DIGEST_LENGTH];
    char *o_key_pad = malloc(sha_block_size);
    char *i_key_pad = malloc(sha_block_size);

    str_xor(o_key_pad, sha_block_size, k, sha_block_size, OPAD);
    str_xor(i_key_pad, sha_block_size, k, sha_block_size, IPAD);

    SHA_CTX ctx;

    // Generating SHA1(i_key_pad||m)
    SHA1_Init(&ctx);

    SHA1_Update(&ctx, i_key_pad, sha_block_size);
    SHA1_Update(&ctx, m, mlen);

    SHA1_Final((unsigned char*) hash, &ctx);

    // Generating SHA1(o_key_pad||SHA1(i_key_pad||m))
    SHA1_Init(&ctx);

    SHA1_Update(&ctx, o_key_pad, sha_block_size);
    SHA1_Update(&ctx, (unsigned char*) hash, SHA_DIGEST_LENGTH);

    SHA1_Final((unsigned char*) hash, &ctx);

    int i;
    for(i = 0; i < SHA_DIGEST_LENGTH; i++)
      res[i] = hash[i];

    free(o_key_pad);
    free(i_key_pad);
}

void hotp(char* output, int klen, char* K, int clen, char* C){
  hmac(output, klen, K, clen, C);
}

void totp(char* output, int klen, char* K, int t){
  char* tc = malloc(sizeof(int));

  int i;
  for(i = 0; i < sizeof(int); i++){
    tc[i] = ((unsigned char*) &t)[sizeof(int) - i - 1];
  }

  hotp(output, klen, K, 4, tc);
}

int getTime(int slotSize){
  int t = time(0);
  int remainder = t % slotSize;

  return t - remainder;

}

void str2hex(char** output, int ilen, char* input){

    int i;
    for(i = 0; i < ilen; i++){
      sprintf(*output + 2*i, "%02x", (unsigned char) input[i]);
    }

}

int main(int argc, char* argv[]){

    // char* a = "abc";
    // char* b = "def";
    char* c = malloc(SHA_DIGEST_LENGTH);
    char *res = malloc(2*SHA_DIGEST_LENGTH);

    char* secret = "0c76f310c1f0bde009dd860a5c09cb118e8c8caa26aa542522b824c7d3296c3331b473553e29c9e4";

    int secretLen = strlen(secret);

    while(1){
      
      int t = getTime(30);
      totp(c, secretLen, secret, t);

      int i;
      for(i = 0; i < SHA_DIGEST_LENGTH; i++){
        sprintf(res + 2*i, "%02x",(unsigned char) c[i]);
      }

      // str2hex(&res, SHA_DIGEST_LENGTH, c);

      printf("%s\n", res);
      sleep(5);
    }

    return 0;
}

File: python/totp.py
#!/usr/bin/env python3
import time
from hashlib import sha1
import uuid

sha_block_size = 64

opad = b'k'* sha_block_size
ipad = b'm'* sha_block_size

def encode(x):
    if type(x) == bytes:
        return x
    else:
        return x.encode()

def decode(x):
    if type(x) == bytes:
        return x.decode()
    else:
        return x

def xor(a, b):

    c = b''

    while len(a) > len(b):
        b = b'\x00' + b

    while len(b) < len(a):
        a = b'\x00' + a

    for i in range(len(a)):
        c += b'%c' % (a[i] ^ b[i])

    return c

def hmac(K, m):

    if len(K) > sha_block_size:
        k = sha1(K).digest()
    else:
        k = K

    while len(k) < sha_block_size:
        k = b'\x00' + k

    o_key_pad = xor(k, opad)
    i_key_pad = xor(k, ipad)

    a = sha1(encode(i_key_pad))
    a.update(encode(m))

    b = sha1(encode(o_key_pad))
    b.update(encode(a.digest()))

    return b.hexdigest()


def hotp(K, C):
    return hmac(K.encode(), C.encode())

def totp(K, TC):
    return hotp(K, int_2_str(TC))

def get_time(slot_size):
    t = int(time.time())
    remainder = t % slot_size

    t -= remainder

    return t

def int_2_str(i):

    a = (i & 0xff000000) >> 24
    b = (i  & 0x00ff0000) >> 16
    c = (i & 0x0000ff00) >> 8
    d = i & 0x000000ff

    return "%c%c%c%c" % (a, b, c, d)

def gen_secret():
    return "%s%s" % (sha1(uuid.uuid4().bytes).hexdigest(), sha1(uuid.uuid4().bytes).hexdigest())


# print(hotp('abc','def'))
secret = gen_secret()
secret = '0c76f310c1f0bde009dd860a5c09cb118e8c8caa26aa542522b824c7d3296c3331b473553e29c9e4'
# print("Secret: %s" % secret)

while 1:
    tc = get_time(30)
    # print("Secret: %s" % secret)
    # print("T: %d" % tc)
    print(totp(secret, tc))
    time.sleep(5)
